### PR TITLE
Simplify Component to ComponentInfo forwarding

### LINKED ISSUE #20266: Expensive Check Component::hasComponentInfo
`Component::hasComponentInfo` is an important part of the compatibility mode we have in the transition between `Instrument 1.0` and `Instrument 2.0`. 

At the time of writing the implementation looks like this:

```cpp
bool Component::hasComponentInfo() const {
  const IComponent *root = m_base;
  while (auto parent = root->getBareParent())
    root = parent;
  auto instrument = dynamic_cast<const Instrument *>(root);
  return m_map->hasComponentInfo(instrument);
}
```
see [here](https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/src/Instrument/Component.cpp#L310-L316)

Which seems like an awfully expensive way of establishing whether we can forward to a `ComponentInfo` or not as it involves a **n** virtual function calls to `getBareParent()` in a loop, as well as a final `dynamic_cast`. The purpose of this seems to be to verify root `Component` of the present `Component` matches the instrument held by the `ParameterMap` instance held by the present `Component`.. before then yielding the `ComponentInfo` from the held `ParameterMap`.

Since we are now starting to use this check in a lot of [places](https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/src/Instrument/Component.cpp#L338), it would be nice to simplify this somehow. Understandably we are still going to have to go via the `ParameterMap` but is the instrument check really necessary?

### PR DESCRIPTION
**Pull request by OwenArnold**
Fixes #20266 contains a description of the problem

**WIP**

Description of fix approach

1. `ParameterMap` can only be associated with a neutronic Instrument (ParameterMap::setInstrument) because `ExperimentInfo` only created for neturonic Instruments
1. Component etc, can therefore reason that their associated ParameterMap has the neutronic `ComponentInfo` attached.
1. `Component` NOT used in anything via physical instrument, so no danger of getting the wrong `ComponentInfo` and attributed when doing things via `Component`

**Discussion Points**

This makes position access x2 faster, so for those bits of the codebase not ported to use `DetectorInfo`/`ComponentInfo` we get an automatic speedup. Also, I think this does make the code more readable and makes client code simpler too. However I have some concerns that there are edge cases I have not considered with this, i.e. there is a genuine need to compare instrument pointers beyond the physical instrument case I have put this PR in for.

### CODE REVIEW

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
I do not really understand how you are preventing accessing components/detectors of the physical instrument?

File: Framework/Geometry/src/Instrument.cpp [hunk 1]
```diff
@@ -517,6 +521,8 @@ auto find(T &map, const detid_t key) -> decltype(map.begin()) {
 *  @throw   NotFoundError If no detector is found for the detector ID given
 */
 IDetector_const_sptr Instrument::getDetector(const detid_t &detector_id) const {
+  if (m_isPhysicalInstrument)
+    throw std::runtime_error("Cannot fetch detector from physical instrument");
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply Framework/Geometry/src/Instrument.cpp]: Correct - I think this is the main point of discussion, and the reason I wanted a second pair of eyes on this. The position information for the detector could give the wrong result for a physicalInstrument.. However, we could also have issues if someone did.

```c++
instrument.getPhyisicalInstrument().getComponentByName("source").getPos()
```
And that is not being tested for.

**Comment by OwenArnold** [comment]
The main weakness of the code applied thus far is exactly as you say - "I do not really understand how you are preventing accessing components/detectors of the physical instrument?"

Since the public `Instrument` API still allows you to extract a physical `Instrument`, future developers could become unstuck if they then use the `IComponent` geometry access methods.

**Comment by SimonHeybrock** [comment]
@OwenArnold Similarely, how is the physical instrument being parsed by `InstrumentView`?

**Comment by OwenArnold** [comment]
@SimonHeybrock that's a good point that I have overlooked. I think `InstrumentVisitor` could give the wrong result because of the presence of the parameter map. I'll investigate but that might make this feature impossible to do correctly.

**Pull request closed without merging**